### ISSUE TITLE
performance of dynamodb-local inside localstack much worse than dynamodb-local outside of localstack.

### ISSUE THREAD

**Issue by dradetsky** [localstack/localstack#3390]
[x] bug report
[x] feature request

This seems like a bug to me, but in any case it's a feature request for better performance.

# Detailed description

I previously wrote this up in https://github.com/localstack/localstack/issues/1205, but nobody seemed to notice. I'm writing it up as its own issue, so that it's clearly distinguished from other problems raised in that issue thread

BEGIN OLD CONTENT

I get very acceptable performance out of either dynamodb-local or dynalite running on bare metal, or dynamodb-local running in docker-compose. However, with dynamodb-local running in localstack running in docker compose, the performance is abysmal.

I'm using localstack in some integration tests I'm writing for a project. FWIW, since the only aws service this project depends on is dynamodb (eks doesn't count), I run localstack with

```
      - SERVICES=dynamodb
```

in the compose file. I ran a particular file of tests against a variety of different dynamodb emulation backends, and timed it with both the jest timer and time(1). Here are my results:
```
* int test perf new
** plain aws inmemory
4.477s, estimated 5s
5.97s user 0.67s system 110% cpu 6.005 total
** plain dynalite inmem
3.935s, estimated 5s
5.23s user 0.50s system 104% cpu 5.488 total
** localstack aws file
37.095s
6.24s user 0.67s system 17% cpu 38.645 total
** localstack aws inmem
35.611s, estimated 38s
5.83s user 0.61s system 17% cpu 37.125 total
** docker aws inmem
6.674s, estimated 36s
5.86s user 0.73s system 80% cpu 8.155 total
```
these mean:
- running dynamodb-local on bare metal with `-inMemory`
- running dynalite on bare metal
- running dynamodb-local in localstack with `-sharedDb`
- running dynamodb-local in localstack with `-inMemory`
- running dynamodb-local in docker-compose with `-inMemory`

For reference, I run localstack by just cloning the repo, ensuring that the repo/image are up-to-date (i.e. `docker-compose pull`), and then just doing docker-compose up. Is performance this bad in that scenario expected? Because it basically means I can't use localstack for anything which does nontrivial work on dynamodb. When I say "nontrivial" I mean: the most intense test in this suite sends about 600 `UpdateItem` requests. Most of them are just a handful of reads & writes. My computer has a 12-core recent intel i7 cpu. It shouldn't be this slow.

the difference between -inMemory, -sharedDb, and dynalite are all pretty trivial (and note that I've also tested e.g. -sharedDb on bare metal & other permutations, I just didn't include them in this comparison). The difference between bare metal and docker-compose is noticeable, but barely (and I would gladly pay the penalty for the convenience).

The difference between localstack and everything else is prohibitive.

END OLD CONTENT

Additionally: I would say the performance of dynamodb-local inside of localstack is prohibitively bad. Like, at my company, we're making a bunch of new microservices that use dynamodb and other aws services. So we want local emulation of dynamodb. And we'd also like to share the same tools & configurations across repos.

I feel like as I try to replicate some of the stuff I've done in this project for other people, I'm going to be saying to them "okay, we'll use localstack for KMS and SQS, but not for Dynamo. For Dynamo, we're just going to run Dynamo-Local in docker separately, because localstack will make it grind to a halt." And that's assuming that nothing else will start exhibiting these problems. My project didn't use KMS or SQS, but maybe I'll be learning about the fun of localstack performance for those services when a colleague brings me his project.

In addition, we need to be able to write integration tests suites and run those tests in CI environments. As it turns out, I found it a whole lot easier to run my tests in a docker container in CI against dynalite than localstack. I never actually managed to get localstack working for this. I'll have to figure this out at some point to help out my co-workers, but i'm starting to wonder whether it's a worthwhile investment. What if, when I finally get it working, it's too slow to use?

It makes me want to start investigating alternatives and finding solutions that don't involve localstack.

**Comment by whummer** [comment]
Potential duplicate of #1205

**Comment by dradetsky** [comment]
@whummer I apologize if I'm cluttering issue space. I thought (and maybe I misread; I've been very busy) that #1205 described an issue that people had successfully resolved by switching to using `-inMemory`. So I thought that maybe nobody was taking my report seriously because, even though I had tested and confirmed that the problem was definitely not dynamodb-local (vs dynalite), `-inMemory`, or docker, but definitely localstack, it was part of an issue that people considered fundamentally resolved (because the original problem of the original reporter had been resolved).

So if I did screw that up feel free to close this in favor of #1205.
